prov/rstream/src/rstream_ep.c

```diff
@@ -0,0 +1,343 @@
+        ret = fi_ep_bind(rstream_ep->ep_fd, &rstream_eq->eq_fd->fid, flags);
+        /* latest bind will always be associated with next cm xchg
+        given rsockets impl. (aka bind then either connect/accept */
+        rstream_eq->bound_ep = rstream_ep;
+	    break;
+    default:
+        FI_WARN(&rstream_prov, FI_LOG_EP_CTRL, "invalid fid class\n");
+        ret = -FI_EINVAL;
+        break;
+    }
+    return ret;
+}
+
+static int rstream_ep_ctrl(struct fid *fid, int command, void *arg)
+{
+    struct rstream_ep *rstream_ep;
+    int ret;
+    rstream_ep = container_of(fid, struct rstream_ep, util_ep.ep_fid.fid);
+
+    switch (command) {
+    case FI_ENABLE:
+        ret = fi_enable(rstream_ep->ep_fd);
+        break;
+    case FI_GETWAIT: /* need to have one cq->fd */
+        ret = fi_control(&rstream_ep->send_cq->fid, FI_GETWAIT, arg);
+        break;
+    default:
+        return -FI_ENOSYS;
+    }
+    return ret;
+}
+
+static struct fi_ops rstream_ep_fi_ops = {
+    .size = sizeof(struct fi_ops),
+    .close = rstream_ep_close,
+    .bind = rstream_ep_bind,
+    .control = rstream_ep_ctrl,
+    .ops_open = fi_no_ops_open,
+};
+
+static struct fi_ops_ep rstream_ops_ep = {
+    .size = sizeof(struct fi_ops_ep),
+    .cancel = fi_no_cancel,
+    .getopt = fi_no_getopt,
+    .setopt = fi_no_setopt,
+    .tx_ctx = fi_no_tx_ctx,
+    .rx_ctx = fi_no_rx_ctx,
+    .rx_size_left = fi_no_rx_size_left,
+    .tx_size_left = fi_no_tx_size_left,
+};
+
+static int rstream_reg_mrs(struct fid_domain *domain,
+    struct rstream_lmr_data *lmr)
+{
+    int ret;
+    uint64_t rx_meta_data_offset = 0;
+    uint32_t full_mr_size = RSTREAM_MAX_MR_SEG_SIZE * 2;
+
+    if (RSTREAM_USING_IWARP)
+        rx_meta_data_offset = RSTREAM_IWARP_DATA_SIZE * RSTREAM_MAX_RX_OPS;
+
+    full_mr_size = full_mr_size + rx_meta_data_offset;
+    lmr->base_addr = malloc(full_mr_size);
+
+    ret = fi_mr_reg(domain, lmr->base_addr, full_mr_size,
+        FI_READ | FI_WRITE | FI_REMOTE_READ | FI_REMOTE_WRITE,
+        0, 0, 0, &lmr->mr, NULL);
+    if (ret)
+        return ret;
+
+    lmr->ldesc = fi_mr_desc(lmr->mr);
+    lmr->rkey = fi_mr_key(lmr->mr);
+    lmr->tx.data_start = (char *)lmr->base_addr;
+    lmr->tx.start_offset = 0;
+    lmr->tx.size = RSTREAM_MAX_MR_SEG_SIZE;
+    lmr->tx.end_offset = 0;
+    lmr->tx.avail_size = lmr->tx.size;
+    lmr->rx.data_start = (char *)lmr->tx.data_start +
+        RSTREAM_MAX_MR_SEG_SIZE + rx_meta_data_offset;
+    lmr->rx.start_offset = 0;
+    lmr->rx.end_offset = 0;
+    lmr->rx.size = RSTREAM_MAX_MR_SEG_SIZE;
+    lmr->rx.avail_size = 0;
+    lmr->recv_buffer_offset = 0;
+    return ret;
+}
+
+static int rstream_cq_init(struct fid_domain *domain, struct rstream_ep *rep)
+{
+    int ret;
+    struct fi_cq_attr attr;
+
+    memset(&attr, 0, sizeof(attr));
+    attr.format = FI_CQ_FORMAT_DATA;
+    attr.wait_obj = FI_WAIT_FD;
+    attr.size = RSTREAM_QP_SIZE;
```

Comment:
We want to allow for adjustable QP size.  Please see the original rsocket code for what options will need to be supported.  This implementation must be compatible with the existing code, particularly since java is going through the process of adding support for it.